File: maze.r
# Maze generator in R
# Joe Wingbermuehle 2013-10-03

# Function to create an empty maze array of width by height.
init_maze <- function(width, height) {
   m <- rep(0, times = width * height)
   dim(m) <- c(width, height)
   m[1, 1:height] <- 1
   m[width, 1:height] <- 1
   m[1:width, 1] <- 1
   m[1:width, height] <- 1
   m
}

# Display a maze.
show_maze <- function(m) {
   for (y in 1:dim(m)[2]) {
      str <- ""
      for (x in 1:dim(m)[1]) {
         if(m[x, y] == 1) {
            str <- paste(str, "[]", sep = '')
         } else {
            str <- paste(str, "  ", sep = '')
         }
      }
      print(str)
   }
}

# Carve a maze starting at x, y.
carve_maze <- function(m, x, y) {
   m[x, y] <- 1
   d <- floor(3 * runif(1))
   xdirs <- c(1, -1, 0, 0)
   ydirs <- c(0, 0, 1, -1)
   for (i in 0:3) {
      j <- 1 + (i + d) %% 4
      dx <- xdirs[j]
      dy <- ydirs[j]
      nx1 <- x + dx
      nx2 <- nx1 + dx
      ny1 <- y + dy
      ny2 <- ny1 + dy
      if (m[nx1, ny1] == 0 && m[nx2, ny2] == 0) {
         m[nx1, ny1] <- 1
         m <- carve_maze(m, nx2, ny2)
      }
   }
   m
}

# Generate and display a random maze.
# Note that the width and height of the maze must be odd.
m <- init_maze(37, 23)
m <- carve_maze(m, 3, 3)
show_maze(m)

File: maze.lua
-- Maze generator in Lua
-- Joe Wingbermuehle 2013-10-06

-- Create an empty maze array.
function init_maze(width, height)
   local result = {}
   for y = 0, height - 1 do
      for x = 0, width - 1 do
         result[y * width + x] = 1
      end
      result[y * width + 0] = 0
      result[y * width + width - 1] = 0
   end
   for x = 0, width - 1 do
      result[0 * width + x] = 0
      result[(height - 1) * width + x] = 0
   end
   return result
end

-- Show a maze.
function show_maze(maze, width, height)
   for y = 0, height - 1 do
      for x = 0, width - 1 do
         if maze[y * width + x] == 0 then
            io.write("  ")
         else
            io.write("[]")
         end
      end
      io.write("\n")
   end
end

-- Carve the maze starting at x, y.
function carve_maze(maze, width, height, x, y)
   local r = math.random(0, 3)
   maze[y * width + x] = 0
   for i = 0, 3 do
      local d = (i + r) % 4
      local dx = 0
      local dy = 0
      if d == 0 then
         dx = 1
      elseif d == 1 then
         dx = -1
      elseif d == 2 then
         dy = 1
      else
         dy = -1
      end
      local nx = x + dx
      local ny = y + dy
      local nx2 = nx + dx
      local ny2 = ny + dy
      if maze[ny * width + nx] == 1 then
         if maze[ny2 * width + nx2] == 1 then
            maze[ny * width + nx] = 0
            carve_maze(maze, width, height, nx2, ny2)
         end
      end
   end
end

-- The size of the maze (must be odd).
width = 39
height = 23

-- Initialize random number generator
math.randomseed(os.time())

-- Generate and display a random maze.
maze = init_maze(width, height)
carve_maze(maze, width, height, 2, 2)
maze[width + 2] = 0
maze[(height - 2) * width + width - 3] = 0
show_maze(maze, width, height)

